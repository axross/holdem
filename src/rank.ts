/**
 * An enum type that expresses a rank part of a card.
 */
export enum Rank {
  Ace = 0,
  Deuce = 1,
  Trey = 2,
  Four = 3,
  Five = 4,
  Six = 5,
  Seven = 6,
  Eight = 7,
  Nine = 8,
  Ten = 9,
  Jack = 10,
  Queen = 11,
  King = 12,
}

/**
 * A string expression of a rank. It's always 1-character length. You can combine this with a SuitChar to build a CardString.
 *
 * @example
 * ```
 * const rankChar: RankChar = "A";
 * const suitChar: SuitChar = "s";
 * const cardString: CardString = `${rankChar}${suitChar}`;
 * ```
 */
export type RankChar =
  | "A"
  | "2"
  | "3"
  | "4"
  | "5"
  | "6"
  | "7"
  | "8"
  | "9"
  | "T"
  | "J"
  | "Q"
  | "K";

/**
 * A utility function set for ranks.
 */
export const RankUtils = Object.freeze({
  /**
   * Parses a char (= 1-charactor-length string) into a Rank.
   *
   * @example
   * ```ts
   * RankUtils.parse("A") === Rank.Ace;
   * RankUtils.parse("T") === Rank.Ten;
   * RankUtils.parse("5") === Rank.Five;
   * ```
   *
   * Only `"A"`, `"K"`, `"Q"`, `"J"`, `"T"`, `"9"`, `"8"`, `"7"`, `"6"`, `"5"`, `"4"`, `"3"` or `"2"` is acceptable.
   *
   * @example
   * ```ts
   * RankUtils.parse("a");  // => Error: "a" is not a valid string value for RankUtils.parse().
   * ```
   *
   * For consistency reason, it needs to be a string even for number-based ranks (e.g. `"4"`).
   *
   * @example
   * ```ts
   * RankUtils.parse("4") === Rank.Four;
   * RankUtils.parse(4);  // Error: 4 is not a valid string value for RankUtils.parse().
   * ```
   */
  parse(char: string): Rank {
    switch (char) {
      case "A":
        return Rank.Ace;
      case "K":
        return Rank.King;
      case "Q":
        return Rank.Queen;
      case "J":
        return Rank.Jack;
      case "T":
        return Rank.Ten;
      case "9":
        return Rank.Nine;
      case "8":
        return Rank.Eight;
      case "7":
        return Rank.Seven;
      case "6":
        return Rank.Six;
      case "5":
        return Rank.Five;
      case "4":
        return Rank.Four;
      case "3":
        return Rank.Trey;
      case "2":
        return Rank.Deuce;
      default:
        throw new Error(
          `${JSON.stringify(char)} is not a valid value for RankUtils.parse().`
        );
    }
  },

  /**
   * Returns power of a Rank.
   *
   * @example
   * ```ts
   * RankUtils.power(Rank.Ace) > RankUtils.power(Rank.King);    // => true
   * RankUtils.power(Rank.Ace) > RankUtils.power(Rank.Deuce);   // => true
   * RankUtils.power(Rank.Trey) > RankUtils.power(Rank.Deuce);  // => true
   * ```
   */
  power(rank: Rank): number {
    switch (rank) {
      case Rank.Ace:
        return 12;
      case Rank.King:
        return 11;
      case Rank.Queen:
        return 10;
      case Rank.Jack:
        return 9;
      case Rank.Ten:
        return 8;
      case Rank.Nine:
        return 7;
      case Rank.Eight:
        return 6;
      case Rank.Seven:
        return 5;
      case Rank.Six:
        return 4;
      case Rank.Five:
        return 3;
      case Rank.Four:
        return 2;
      case Rank.Trey:
        return 1;
      case Rank.Deuce:
        return 0;
    }
  },

  /**
   * Compares two ranks in index order and returns integer compatible with Array#sort().
   *
   * @example
   * ```
   * RankUtils.compare(Rank.Ace, Rank.Deuce);  // => negative integer
   * RankUtils.compare(Rank.Deuce, Rank.Ace);  // => positive integer
   * RankUtils.compare(Rank.Ace, Rank.Ace);  // => 0
   * ```
   */
  compare(a: Rank, b: Rank): number {
    return a - b;
  },

  /**
   * Compares two ranks in power order and returns integer compatible with Array#sort().
   *
   * @example
   * ```
   * RankUtils.compare(Rank.Ace, Rank.King);  // => negative integer
   * RankUtils.compare(Rank.King, Rank.Ace);  // => positive integer
   * RankUtils.compare(Rank.Ace, Rank.Ace);  // => 0
   * RankUtils.compare(Rank.Ace, Rank.Deuce);  // => positive integer
   * ```
   */
  comparePower(a: Rank, b: Rank): number {
    return RankUtils.power(b) - RankUtils.power(a);
  },

  /**
   * Stringifies a Rank.
   *
   * @example
   * ```ts
   * RankUtils.format(Rank.Ace) === "A";
   * RankUtils.format(Rank.Ten) === "T";
   * RankUtils.format(Rank.Five) === "5";
   * ```
   *
   * You can utilize this function to build a CardString.
   *
   * @example
   * ```ts
   * const aceChar = RankUtils.format(Rank.Ace);
   * const spadeChar = RankUtils.format(Suit.Spade);
   * const aceOfSpade: CardString = `${aceChar}${spadeChar}`;
   * ```
   */
  format(rank: Rank): RankChar {
    switch (rank) {
      case Rank.Ace:
        return "A";
      case Rank.King:
        return "K";
      case Rank.Queen:
        return "Q";
      case Rank.Jack:
        return "J";
      case Rank.Ten:
        return "T";
      case Rank.Nine:
        return "9";
      case Rank.Eight:
        return "8";
      case Rank.Seven:
        return "7";
      case Rank.Six:
        return "6";
      case Rank.Five:
        return "5";
      case Rank.Four:
        return "4";
      case Rank.Trey:
        return "3";
      case Rank.Deuce:
        return "2";
    }
  },
});
