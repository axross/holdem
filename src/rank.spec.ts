import { Rank, RankUtils } from "./rank";

describe("RankUtils.parse()", () => {
  it('parses "A" into Rank.Ace', () => {
    expect(RankUtils.parse("A")).toBe(Rank.Ace);
  });

  it('parses "K" into Rank.King', () => {
    expect(RankUtils.parse("K")).toBe(Rank.King);
  });

  it('parses "Q" into Rank.Queen', () => {
    expect(RankUtils.parse("Q")).toBe(Rank.Queen);
  });

  it('parses "J" into Rank.Jack', () => {
    expect(RankUtils.parse("J")).toBe(Rank.Jack);
  });

  it('parses "T" into Rank.Ten', () => {
    expect(RankUtils.parse("T")).toBe(Rank.Ten);
  });

  it('parses "9" into Rank.Nine', () => {
    expect(RankUtils.parse("9")).toBe(Rank.Nine);
  });

  it('parses "8" into Rank.Eight', () => {
    expect(RankUtils.parse("8")).toBe(Rank.Eight);
  });

  it('parses "7" into Rank.Seven', () => {
    expect(RankUtils.parse("7")).toBe(Rank.Seven);
  });

  it('parses "6" into Rank.Six', () => {
    expect(RankUtils.parse("6")).toBe(Rank.Six);
  });

  it('parses "5" into Rank.Five', () => {
    expect(RankUtils.parse("5")).toBe(Rank.Five);
  });

  it('parses "4" into Rank.Four', () => {
    expect(RankUtils.parse("4")).toBe(Rank.Four);
  });

  it('parses "3" into Rank.Trey', () => {
    expect(RankUtils.parse("3")).toBe(Rank.Trey);
  });

  it('parses "2" into Rank.Deuce', () => {
    expect(RankUtils.parse("2")).toBe(Rank.Deuce);
  });

  it('throws an error because "a" is not a valid string', () => {
    expect(() => RankUtils.parse("a")).toThrowErrorMatchingSnapshot();
  });

  it('throws an error because "" is not a valid string', () => {
    expect(() => RankUtils.parse("")).toThrowErrorMatchingSnapshot();
  });
});

describe("RankUtils.power()", () => {
  it("returns a power index of a Rank", () => {
    expect(RankUtils.power(Rank.Ace)).toBeGreaterThan(
      RankUtils.power(Rank.King)
    );
    expect(RankUtils.power(Rank.King)).toBeGreaterThan(
      RankUtils.power(Rank.Queen)
    );
    expect(RankUtils.power(Rank.Queen)).toBeGreaterThan(
      RankUtils.power(Rank.Jack)
    );
    expect(RankUtils.power(Rank.Jack)).toBeGreaterThan(
      RankUtils.power(Rank.Ten)
    );
    expect(RankUtils.power(Rank.Ten)).toBeGreaterThan(
      RankUtils.power(Rank.Nine)
    );
    expect(RankUtils.power(Rank.Nine)).toBeGreaterThan(
      RankUtils.power(Rank.Eight)
    );
    expect(RankUtils.power(Rank.Eight)).toBeGreaterThan(
      RankUtils.power(Rank.Seven)
    );
    expect(RankUtils.power(Rank.Seven)).toBeGreaterThan(
      RankUtils.power(Rank.Six)
    );
    expect(RankUtils.power(Rank.Six)).toBeGreaterThan(
      RankUtils.power(Rank.Five)
    );
    expect(RankUtils.power(Rank.Five)).toBeGreaterThan(
      RankUtils.power(Rank.Four)
    );
    expect(RankUtils.power(Rank.Four)).toBeGreaterThan(
      RankUtils.power(Rank.Trey)
    );
    expect(RankUtils.power(Rank.Trey)).toBeGreaterThan(
      RankUtils.power(Rank.Deuce)
    );
  });
});

describe("RankUtils.compare()", () => {
  it("returns negative integer as comparison result of Rank.Ace and Rank.Deuce", () => {
    expect(RankUtils.compare(Rank.Ace, Rank.Deuce)).toBeLessThan(0);
  });

  it("returns positive integer as comparison result of Rank.Four and Rank.Ace", () => {
    expect(RankUtils.compare(Rank.Four, Rank.Ace)).toBeGreaterThan(0);
  });

  it("returns 0 as comparison result of Rank.Six and Rank.Six", () => {
    expect(RankUtils.compare(Rank.Six, Rank.Six)).toBe(0);
  });
});

describe("RankUtils.comparePower()", () => {
  it("returns negative integer as comparison result of Rank.Ace and Rank.King", () => {
    expect(RankUtils.comparePower(Rank.Ace, Rank.King)).toBeLessThan(0);
  });

  it("returns positive integer as comparison result of Rank.Queen and Rank.Five", () => {
    expect(RankUtils.comparePower(Rank.Five, Rank.Eight)).toBeGreaterThan(0);
  });

  it("returns 0 as comparison result of Rank.Ten and Rank.Ten", () => {
    expect(RankUtils.comparePower(Rank.Ten, Rank.Ten)).toBe(0);
  });

  it("returns negative integer as comparison result of Rank.King and Rank.Deuce", () => {
    expect(RankUtils.comparePower(Rank.King, Rank.Deuce)).toBeLessThan(0);
  });
});

describe("RankUtils.format()", () => {
  it('stringifies Rank.Ace into "A"', () => {
    expect(RankUtils.format(Rank.Ace)).toBe("A");
  });

  it('stringifies Rank.King into "K"', () => {
    expect(RankUtils.format(Rank.King)).toBe("K");
  });

  it('stringifies Rank.Queen into "Q"', () => {
    expect(RankUtils.format(Rank.Queen)).toBe("Q");
  });

  it('stringifies Rank.Jack into "J"', () => {
    expect(RankUtils.format(Rank.Jack)).toBe("J");
  });

  it('stringifies Rank.Ten into "T"', () => {
    expect(RankUtils.format(Rank.Ten)).toBe("T");
  });

  it('stringifies Rank.Nine into "9"', () => {
    expect(RankUtils.format(Rank.Nine)).toBe("9");
  });

  it('stringifies Rank.Eight into "8"', () => {
    expect(RankUtils.format(Rank.Eight)).toBe("8");
  });

  it('stringifies Rank.Seven into "7"', () => {
    expect(RankUtils.format(Rank.Seven)).toBe("7");
  });

  it('stringifies Rank.Six into "6"', () => {
    expect(RankUtils.format(Rank.Six)).toBe("6");
  });

  it('stringifies Rank.Five into "5"', () => {
    expect(RankUtils.format(Rank.Five)).toBe("5");
  });

  it('stringifies Rank.Four into "4"', () => {
    expect(RankUtils.format(Rank.Four)).toBe("4");
  });

  it('stringifies Rank.Trey into "3"', () => {
    expect(RankUtils.format(Rank.Trey)).toBe("3");
  });

  it('stringifies Rank.Deuce into "2"', () => {
    expect(RankUtils.format(Rank.Deuce)).toBe("2");
  });
});
