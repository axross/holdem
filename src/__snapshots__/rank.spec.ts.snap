// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RankUtils.parse() throws an error because "" is not a valid string 1`] = `""" is not a valid value for RankUtils.parse()."`;

exports[`RankUtils.parse() throws an error because "a" is not a valid string 1`] = `""a" is not a valid value for RankUtils.parse()."`;
